contracts/PartyBid.sol

```diff
@@ -0,0 +1,701 @@
+        emit Finalized(partyStatus, totalSpent, _ethFee, totalContributedToParty);
+    }
+
+    // ======== External: Claim =========
+
+    /**
+     * @notice Claim the tokens and excess ETH owed
+     * to a single contributor after the auction has ended
+     * @dev Emits a Claimed event upon success
+     * callable by anyone (doesn't have to be the contributor)
+     * @param _contributor the address of the contributor
+     */
+    function claim(address _contributor) external nonReentrant {
+        // ensure auction has finalized
+        require(
+            partyStatus != PartyStatus.AUCTION_ACTIVE,
+            "PartyBid::claim: auction not finalized"
+        );
+        // ensure contributor submitted some ETH
+        require(
+            totalContributed[_contributor] != 0,
+            "PartyBid::claim: not a contributor"
+        );
+        // ensure the contributor hasn't already claimed
+        require(
+            !claimed[_contributor],
+            "PartyBid::claim: contribution already claimed"
+        );
+        // mark the contribution as claimed
+        claimed[_contributor] = true;
+        // calculate the amount of fractional NFT tokens owed to the user
+        // based on how much ETH they contributed towards the auction,
+        // and the amount of excess ETH owed to the user
+        (uint256 _tokenAmount, uint256 _ethAmount) =
+            getClaimAmounts(_contributor);
+        // transfer tokens to contributor for their portion of ETH used
+        _transferTokens(_contributor, _tokenAmount);
+        // if there is excess ETH, send it back to the contributor
+        _transferETHOrWETH(_contributor, _ethAmount);
+        emit Claimed(
+            _contributor,
+            totalContributed[_contributor],
+            _ethAmount,
+            _tokenAmount
+        );
+    }
+
+    // ======== External: Emergency Escape Hatches (PartyDAO Multisig Only) =========
+
+    /**
+     * @notice Escape hatch: in case of emergency,
+     * PartyDAO can use emergencyWithdrawEth to withdraw
+     * ETH stuck in the contract
+     */
+    function emergencyWithdrawEth(uint256 _value)
+        external
+        onlyPartyDAO
+    {
+        _transferETHOrWETH(partyDAOMultisig, _value);
+    }
+
+    /**
+     * @notice Escape hatch: in case of emergency,
+     * PartyDAO can use emergencyCall to call an external contract
+     * (e.g. to withdraw a stuck NFT or stuck ERC-20s)
+     */
+    function emergencyCall(address _contract, bytes memory _calldata)
+        external
+        onlyPartyDAO
+        returns (bool _success, bytes memory _returnData)
+    {
+        (_success, _returnData) = _contract.call(_calldata);
+        require(_success, string(_returnData));
+    }
+
+    /**
+     * @notice Escape hatch: in case of emergency,
+     * PartyDAO can force the PartyBid to finalize with status LOST
+     * (e.g. if finalize is not callable)
+     */
+    function emergencyForceLost()
```

Comment:
yeah this would be the more pure approach, the emergency functions are meant to help us recover in the case that something horrible and unexpected happens on a super high value party, because the PartyBid is interacting with a lot of different contracts and standards under the hood (ERC-721, auction houses, market wrappers) so there is room for unknown-unknowns. but the intention is essentially to that we should never need to call any of the above